### ISSUE TITLE
[database corruption] strange exception after testing site stability

### ISSUE THREAD

**Issue by RarogCmex** [robiso/wondercms#200]
Site https://imperium.org.ru
#0 /srv/wondercms-imperium.org.ru/index.php(92): Wcms->getDb()
#1 /srv/wondercms-imperium.org.ru/index.php(14): Wcms->__construct()
#2 {main}
  thrown in /srv/wondercms-imperium.org.ru/index.php on line 539" while reading response header from upstream, client: 192.168.88.210, server: imperium.org.ru, request: "GET / HTTP/2.0", upstream: "fastcgi://unix:/run/imperium-php-fpm.sock:", host: "imperium.org.ru"
2020/05/13 00:22:12 [error] 25915#25915: *1 FastCGI sent in stderr: "PHP message: PHP Fatal error:  Uncaught TypeError: Return value of Wcms::getDb() must be an instance of stdClass, null returned in /srv/wondercms-imperium.org.ru/index.php:539
Stack trace:
#0 /srv/wondercms-imperium.org.ru/index.php(92): Wcms->getDb()
#1 /srv/wondercms-imperium.org.ru/index.php(14): Wcms->__construct()
#2 {main}
  thrown in /srv/wondercms-imperium.org.ru/index.php on line 539" while reading response header from upstream, client: 192.168.88.210, server: imperium.org.ru, request: "GET / HTTP/2.0", upstream: "fastcgi://unix:/run/imperium-php-fpm.sock:", host: "imperium.org.ru"
2020/05/13 00:24:10 [error] 25915#25915: *1 FastCGI sent in stderr: "PHP message: PHP Fatal error:  Uncaught TypeError: Return value of Wcms::getDb() must be an instance of stdClass, null returned in /srv/wondercms-imperium.org.ru/index.php:539
Stack trace:
#0 /srv/wondercms-imperium.org.ru/index.php(92): Wcms->getDb()
#1 /srv/wondercms-imperium.org.ru/index.php(14): Wcms->__construct()
#2 {main}
  thrown in /srv/wondercms-imperium.org.ru/index.php on line 539" while reading response header from upstream, client: 192.168.88.210, server: imperium.org.ru, request: "GET / HTTP/2.0", upstream: "fastcgi://unix:/run/imperium-php-fpm.sock:", host: "imperium.org.ru"

Now my site is down. How to fix it?

**Comment by RarogCmex** [comment]
@robiso should I place web site snapshot?
nginx config:
`
user melkor; #Special user for security. Make sure you changed permissions of /var/lib/nginx/tmp/
worker_processes 4;
error_log /var/log/nginx/error_log info;
pid /var/run/nginx.pid;
events {
 worker_connections 512;
 use epoll;
}
http {
 include mime.types; #TODO: add more mimetypes
 default_type application/octet-stream;
##Timeouts and limits
 client_header_timeout 10m;
 client_body_timeout 10m;
 send_timeout 10m;
 client_max_body_size 100m;
##Compression
###Old style compression — gzip. However, can be overriden for some locations
 gzip off; #Disabled because there are brotli replacement
###New type of compression algoritm — brotli
###See https://github.com/google/ngx_brotli
###Also brotli --help
 brotli on; #Enables
 brotli_static on; #Enables precompressed ones like jquery.js.br
 brotli_comp_level 4; #Level of compression; By default 6; Range from 0 to 11;
#brotli_types *; #All compressed, enable only for the desired sites
 brotli_min_length 128; #Do not use compression for files smaller that 128 bytes; By default 20
 brotli_window 512k; #Default. Can be overriden for some locations or servers. Acceptable values are 1k, 2k, 4k, 8k, 16k, 32k, 64k, 128k, 25>
server {
  listen 80;
  server_name imperium.org.ru;
  return 301 https://$host$request_uri;
 }
 server {
  listen 443 ssl http2;
  ssl_certificate /etc/letsencrypt/live/imperium.org.ru/fullchain.pem; # managed by Certbot
  ssl_certificate_key /etc/letsencrypt/live/imperium.org.ru/privkey.pem; # managed by Certbot
  include /etc/letsencrypt/options-ssl-nginx.conf; # managed by Certbot
  ssl_dhparam /etc/letsencrypt/ssl-dhparams.pem; # managed by Certbot
  server_name imperium.org.ru;
  root /srv/wondercms-imperium.org.ru;
  index index.php;
  include fastcgi.conf;
  brotli_types *; #All compressed
  autoindex off; #Prevent directory listing
  # rewrite url to make it pretty
  location / {
   if (!-e $request_filename) {
   rewrite ^/(.+)$ /index.php?page=$1 last; }
  }
  # prevent access to database.js
  location ~ database.js {
   return 403;
  }
  location ~ cache.json {
   return 403;
  location ~ cache.json {
   return 403;
  }
  # use php-fpm for dealing with php files
  location ~ \.php(/|$) {
   # using unix socket
   try_files $uri =404;
   fastcgi_pass unix:/run/imperium-php-fpm.sock;
  }
}
`

**Comment by RarogCmex** [comment]
```
RarogCmexExp ~ # LC_ALL=C diff /srv/wondercms-imperium.org.ru/data/database.js /mnt/data/Ds.Backups/beta.1/localhost/srv/wondercms-imperium.org.ru/data/database.js 
Binary files /srv/wondercms-imperium.org.ru/data/database.js and /mnt/data/Ds.Backups/beta.1/localhost/srv/wondercms-imperium.org.ru/data/database.js differ

```

**Comment by RarogCmex** [comment]
Seems like database.js corrupted

**Comment by RarogCmex** [comment]
@robiso send compressed site archive with database to you r@robiso.com

**Comment by RarogCmex** [comment]
Database is recovered from backup, site now working.
May be I have just bad webserver, because I created ones from semi-broken notebook 2010 yearborn :(
Accident is happened when I experiment with pm.max_children = 6 in php settings. I tries to increase it up to 30 to see memory consumption.

**Comment by robiso** [comment]
@RarogCmex, thanks for the report.
Is there any chance you could create a backup of your website again and try to reproduce the same case?

**Comment by RarogCmex** [comment]
@robiso I'll try that, but may be it's hardware error.

**Comment by robiso** [comment]
@RarogCmex looking forward and intrigued to see what the results will be.

**Comment by RarogCmex** [comment]
@robiso It seems like hardware error. I can't reproduce it.

**Comment by RarogCmex** [comment]
Should ticket be closed?

**Comment by robiso** [comment]
@RarogCmex thanks for getting back to me on a such short notice. I'll close the issue, but please feel to re-open it if you run into any issues.

If you're interested in something a bit more fresh, I've just published an pre-release version of WonderCMS 3.0.8 with some interesting improvements: 
- https://github.com/robiso/wondercms/tree/experimental3.0.8

Let me know if you want to run it for a spin, should work even faster.
